File: ruby/unique_paths.rb
# @param {Integer} m
# @param {Integer} n
# @return {Integer}
def unique_paths(m, n)
    unique_paths_helper(m, n, {})
end

def unique_paths_helper(m, n, cache)
    return 1 if (n == 1 || m == 1)
    
    last_paths_count_from_left = cache[[m, n - 1]] || unique_paths_helper(m, n - 1, cache)
    last_paths_count_from_top = cache[[m - 1, n]] || unique_paths_helper(m - 1, n, cache)
    
    result = last_paths_count_from_left + last_paths_count_from_top
    cache[[m, n]] = result
    result
end

File: javascript/unique_paths.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function(m, n) {
    return dp(m, n, {});
};

function dp(m, n, cache) {
    if (m === 0 || n === 0) { return 0; }
    if (m === 1 || n === 1) { return 1; }

    var cacheKeym = (m - 1) + ',' + n;
    var cacheKeyn = m + ',' + (n - 1);

    cache[cacheKeym] = cache[cacheKeym] || dp(m - 1, n, cache);
    cache[cacheKeyn] = cache[cacheKeyn] || dp(m, n - 1, cache);

    return cache[cacheKeym] + cache[cacheKeyn];
}
